### PR TITLE
Migrate template backend to new api

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/core/dev_api/openvino/runtime/itensor.hpp [hunk 1]
```diff
@@ -23,6 +23,13 @@ class OPENVINO_API ITensor : public std::enable_shared_from_this<ITensor> {
      */
     virtual void set_shape(ov::Shape shape) = 0;
 
+    /**
+     * @brief Set new element type for tensor
+     * @note Memory allocation may happen
+     * @param shape A new shape
+     */
+    virtual void set_element_type(ov::element::Type element_type);
```
Comment (ilya-lavrenov) [inline src/core/dev_api/openvino/runtime/itensor.hpp:31]: better to avoid such API

Comment (ilyachur) [reply src/core/dev_api/openvino/runtime/itensor.hpp:31]: I will remove it

File: src/plugins/template/backend/ops/ctc_loss.cpp [hunk 2]
```diff
@@ -2,62 +2,59 @@
...
-    using T1 = typename ngraph::element_type_traits<t1>::value_type;
-    using T2 = typename ngraph::element_type_traits<t2>::value_type;
-    ov::reference::CTCLoss<T1, T2>(inputs[0]->get_data_ptr<T1>(),
-                                   inputs[0]->get_shape(),
-                                   inputs[1]->get_data_ptr<T2>(),
-                                   inputs[2]->get_data_ptr<T2>(),
-                                   inputs[3]->get_data_ptr<T2>(),
-                                   inputs[4]->get_data_ptr<T2>(),
+template <
+    ov::element::Type_t t1,
```
Comment (praasz) [inline src/plugins/template/backend/ops/ctc_loss.cpp:24]: It is detail but maybe use ET0, ET1 etc. here?

Comment (ilyachur) [reply src/plugins/template/backend/ops/ctc_loss.cpp:24]: Yes will change in next PR if CI is passed.

File: src/core/reference/src/op/non_max_suppression.cpp [hunk 3]
```diff
@@ -366,35 +366,34 @@ void non_max_suppression(const float* boxes_data,
...
-        outputs[1]->set_element_type(selected_scores_type);
-        outputs[1]->set_shape(Shape{static_cast<size_t>(valid_outputs), 3});
+        // outputs[1].set_element_type(selected_scores_type);
+        outputs[1].set_shape(Shape{static_cast<size_t>(valid_outputs), 3});
     }
 
     if (num_of_outputs >= 3) {
-        outputs[2]->set_element_type(output_type);
-        outputs[2]->set_shape(Shape{1});
+        // outputs[2].set_element_type(output_type);
```
Comment (ilya-lavrenov) [inline src/core/reference/src/op/non_max_suppression.cpp:386]: remove commented code?

File: src/core/reference/include/openvino/reference/adaptive_max_pool.hpp [hunk 4]
```diff
@@ -10,6 +10,7 @@
 
 #include "ngraph/axis_vector.hpp"
 #include "ngraph/shape.hpp"
+#include "openvino/reference/adaptive_avg_pool.hpp"
```
Comment (ilya-lavrenov) [inline src/core/reference/include/openvino/reference/adaptive_max_pool.hpp:13]: do we need `avg_pool` in `max_pool`?

File: src/plugins/template/backend/ops/ctc_loss.cpp (hunk 2 above)
Comment (ilyachur) [reply src/plugins/template/backend/ops/ctc_loss.cpp]: Fixed

File: src/core/reference/src/op/non_max_suppression.cpp (hunk 3 above)
Comment (ilyachur) [reply src/core/reference/src/op/non_max_suppression.cpp]: Fixed

Comment (ilya-lavrenov) [reply src/core/reference/src/op/non_max_suppression.cpp]: we have several other similar places in this file

Comment (ilyachur) [reply src/core/reference/src/op/non_max_suppression.cpp]: Removed all comments in backend ops

File: src/core/src/node.cpp [hunk 5]
```diff
@@ -740,7 +740,8 @@ inline void update_output_tensors(ov::TensorVector& output_values, const ngraph:
             if (!tensor && outputs[i]->get_partial_shape().is_static()) {
                 tensor = ov::Tensor(outputs[i]->get_element_type(), outputs[i]->get_shape());
             }
-            output_values[i] = tensor;
+            output_values[i].set_shape(tensor.get_shape());
+            tensor.copy_to(output_values[i]);
```
Comment (ilya-lavrenov) [inline src/core/src/node.cpp]: should copy_to automatically set_shape?

Comment (ilyachur) [reply src/core/src/node.cpp]: Yes it can, but in case if shapes are the same it won't modify shapes.

**Review by ilya-lavrenov — APPROVED** [review]

File: src/core/reference/src/op/non_max_suppression.cpp [hunk 6]
```diff
@@ -366,35 +366,33 @@ void non_max_suppression(const float* boxes_data,
...
+                        const ov::element::Type selected_scores_type) {
+    // outputs[0].set_element_type(output_type);
+    outputs[0].set_shape(Shape{static_cast<size_t>(valid_outputs), 3});
 
     size_t num_of_outputs = outputs.size();
 
     if (num_of_outputs >= 2) {
-        outputs[1]->set_element_type(selected_scores_type);
-        outputs[1]->set_shape(Shape{static_cast<size_t>(valid_outputs), 3});
+        // outputs[1].set_element_type(selected_scores_type);
```
Comment (ilya-lavrenov) [inline src/core/reference/src/op/non_max_suppression.cpp:381]: set_element_type is still here :)

**Pull request merged**